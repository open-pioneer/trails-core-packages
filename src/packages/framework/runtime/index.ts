// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
export { type PackageIntl } from "./i18n";
export {
    type ApplicationElement,
    type ApplicationElementConstructor,
    type ApplicationProperties,
    type ApplicationConfig,
    type CustomElementOptions,
    type ConfigContext,
    type RawApplicationProperties,
    createCustomElement
} from "./CustomElement";
export * from "./Service";
export * from "./ServiceRegistry";
export * from "./PropertiesRegistry";

export { ApiServiceImpl } from "./builtin-services/ApiServiceImpl";
