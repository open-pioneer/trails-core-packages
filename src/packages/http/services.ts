// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import type { HttpService } from "./index";

export class HttpServiceImpl implements HttpService {
    async fetch(resource: RequestInfo | URL, init?: RequestInit | undefined): Promise<Response> {
        return await globalThis.fetch(resource, init);
    }
}
