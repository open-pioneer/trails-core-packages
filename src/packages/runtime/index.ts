// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
export { type PackageIntl } from "./i18n";
export {
    type ApplicationElement,
    type ApplicationElementConstructor,
    type ApplicationProperties,
    type ApplicationConfig,
    type CustomElementOptions,
    type ConfigContext,
    createCustomElement
} from "./CustomElement";
export * from "./Service";
export {
    type DeclaredService,
    type AssociatedInterfaceName,
    type InterfaceNameForServiceType
} from "./DeclaredService";
