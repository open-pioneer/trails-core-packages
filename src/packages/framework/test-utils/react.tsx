import { createIntl, createIntlCache, IntlShape } from "@formatjs/intl";
import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const contextMethods = useMemo(() => createPackageContextMethods(rest), [rest]);
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, IntlShape> = {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                const cache = createIntlCache();
                return createIntl(
                    {
                        locale,
                        messages: packageMessages
                    },
                    cache
                );
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
}
