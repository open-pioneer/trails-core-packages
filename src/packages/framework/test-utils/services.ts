// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ReferenceMeta, Service, ServiceConstructor } from "@open-pioneer/runtime";
import { createIntl, createIntlCache } from "@formatjs/intl";

/**
 * Options for the {@link createService} function.
 */
export interface CreateServiceOptions<References> {
    /**
     * References provided to the service constructor.
     *
     * @default {}
     */
    references?: PartialServiceReferences<References>;

    /**
     * Properties provided to the service constructor.
     *
     * @default {}
     */
    properties?: Record<string, unknown>;

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * The locale for embedded default messages.
     *
     * See also https://formatjs.io/docs/intl#message-descriptor
     *
     * @default "en"
     */
    defaultMessageLocale?: string;

    /**
     * I18n messages as (messageId, message) entries.
     *
     * @default {}
     */
    messages?: Record<string, string>;
}

type PartialServiceReferences<References> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [referenceName in keyof References]?: References[referenceName] extends any[]
        ? Partial<References[referenceName][number]>[]
        : Partial<References[referenceName]>;
};

/**
 * Creates a new service instance of `clazz`.
 *
 * The options passed into the constructor of `clazz` are taking from the `options` parameter.
 *
 * This function allows the user to define mock- or test objects as service references.
 *
 * Most type-checking features are disabled for convenience, but by default this function's types
 * enforce that a value provided for a reference must match the partial interface of the actually required type.
 *
 * For example, if a service depends on `{ ctx: ApplicationContext }`, this function ensures
 * that the test value for `ctx` implements `Partial<ApplicationContext>`.
 * This can of course be disabled by casting the value explicitly.
 *
 * @param clazz a service class
 * @param options options passed to the service constructor
 * @returns a new instance of the given service class
 */
export async function createService<References extends {}, Interface extends {}>(
    clazz: ServiceConstructor<References, Interface>,
    options?: CreateServiceOptions<References>
): Promise<Service<Interface>> {
    const references = options?.references ?? {};
    const referencesMeta = Object.fromEntries(
        Object.entries(references).map(([referenceName, reference]) => [
            referenceName,
            referenceMeta(referenceName, reference)
        ])
    );

    const locale = options?.locale ?? "en";
    const defaultMessageLocale = options?.defaultMessageLocale ?? "en";
    const messages = options?.messages ?? {};
    const cache = createIntlCache();
    const intl = createIntl(
        {
            locale,
            defaultLocale: defaultMessageLocale,
            messages
        },
        cache
    );

    return new clazz({
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        references: references as any,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        referencesMeta: referencesMeta as any,
        properties: options?.properties ?? {},
        intl
    });
}

function referenceMeta(referenceName: string, reference: unknown): ReferenceMeta | ReferenceMeta[] {
    if (Array.isArray(reference)) {
        return reference.map<ReferenceMeta>((_, index) => ({
            serviceId: `test-utils::${referenceName}-${index}`
        }));
    }
    return {
        serviceId: `test-utils::${referenceName}`
    };
}
