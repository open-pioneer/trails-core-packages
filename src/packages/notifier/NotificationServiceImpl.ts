// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createToaster, CreateToasterProps, CreateToasterReturn } from "@chakra-ui/react";
import type {
    NotificationLevel,
    NotificationOptions,
    NotificationService,
    NotifierProperties,
    SimpleNotificationOptions
} from "./api";
import { Resource, createLogger } from "@open-pioneer/core";
import { ApplicationContext, ServiceOptions } from "@open-pioneer/runtime";
const LOG = createLogger("notifier:NotificationService");

export type ToasterObject = CreateToasterReturn;

export interface ToastMeta {
    closable: boolean;
}

export interface InternalNotificationAPI extends NotificationService {
    // Toasts in here carry `ToastMeta` metadata
    readonly toaster: ToasterObject;
    registerUI(): Resource | undefined;
}

interface References {
    appCtx: ApplicationContext;
}

// 7 days. Needed because there is no "indefinite" timeout for tooltips in current chakra.
const PERSISTENT_TIMEOUT = 7 * 24 * 60 * 60 * 1000;

export class NotificationServiceImpl implements InternalNotificationAPI {
    #uiPresent = false;
    #uiCheckTimeoutId: any; // eslint-disable-line @typescript-eslint/no-explicit-any

    readonly toaster: ToasterObject;

    constructor({ properties }: ServiceOptions<References>) {
        const typedProperties = properties as NotifierProperties;

        this.toaster = createToaster({
            placement: getPlacement(typedProperties.position),
            pauseOnPageIdle: true,
        });

        if (import.meta.env.DEV && !import.meta.env.VITEST) {
            this.#uiCheckTimeoutId = setTimeout(() => {
                this.#checkUiNotification();
                this.#uiCheckTimeoutId = undefined;
            }, 5000);
        }
    }

    destroy() {
        clearTimeout(this.#uiCheckTimeoutId);
        this.#uiCheckTimeoutId = undefined;
    }

    notify(options: NotificationOptions): void {
        this.toaster.create({
            type: options.level ?? "info",
            title: options.title,
            description: options.message,

            // Chakra currently has no concept of persistent toasts (other than type "loading"),
            // so we just use a very long timeout instead.
            // Note: MAX_VALUE or Infinity does not work either (probably too big for some computations made internally).
            duration: options.displayDuration ?? PERSISTENT_TIMEOUT,

            // Additional data for the toast (can be arbitrary)
            meta: {
                closable: true
            } satisfies ToastMeta
        });
    }

    success(options: SimpleNotificationOptions): void {
        this.#sendSimpleNotification("success", options);
    }

    info(options: SimpleNotificationOptions): void {
        this.#sendSimpleNotification("info", options);
    }

    warning(options: SimpleNotificationOptions): void {
        this.#sendSimpleNotification("warning", options);
    }

    error(options: SimpleNotificationOptions): void {
        this.#sendSimpleNotification("error", options);
    }

    #sendSimpleNotification(level: NotificationLevel, options: SimpleNotificationOptions): void {
        if (typeof options === "string") {
            options = { message: options };
        }
        this.notify({ ...options, level });
    }

    closeAll(): void {
        this.toaster.dismiss();
    }

    registerUI(): Resource | undefined {
        // We only support exactly one handler.
        if (this.#uiPresent) {
            LOG.warn(
                "A notifier UI has already been registered; this new component will be ignored.\n" +
                    "The <Notifier /> component has likely been used twice in your application."
            );
            return undefined;
        }

        this.#uiPresent = true;
        let destroyed = false;
        return {
            destroy: () => {
                if (destroyed) {
                    return;
                }

                destroyed = true;
                this.#uiPresent = false;
            }
        };
    }

    #checkUiNotification() {
        if (!this.#uiPresent) {
            LOG.warn(
                `No notifier UI has been registered: notifications will not be visible.\n` +
                    `Make sure that your app contains the <Notifier /> component.`
            );
        }
    }
}

// Map to chakra placement
function getPlacement(
    configPosition: NotifierProperties["position"] = "top-right"
): CreateToasterProps["placement"] {
    switch (configPosition) {
        case "top":
            return "top";
        case "top-left":
            return "top-start";
        case "top-right":
            return "top-end";
        case "bottom":
            return "bottom";
        case "bottom-left":
            return "bottom-start";
        case "bottom-right":
            return "bottom-end";
        default:
            return "top-end";
    }
}
