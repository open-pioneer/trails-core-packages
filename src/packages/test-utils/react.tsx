// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { PackageIntl, Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime-react-support";
import { CustomChakraProvider } from "@open-pioneer/runtime/test-support";
import { FC, ReactNode, useMemo } from "react";
import { createIntl } from "./vanilla";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyService = Service<any>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using Open Pioneer Trails hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 * Currently also wrapped with the CustomChakraProvider to support chakra-ui elements
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const [locale, contextMethods] = useMemo(() => createPackageContextMethods(rest), [rest]);

    return (
        <CustomChakraProvider rootNode={document} appRoot={document.body} locale={locale}>
            <InternalPackageContext.Provider value={contextMethods}>
                {children}
            </InternalPackageContext.Provider>
        </CustomChakraProvider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): [locale: string, PackageContextMethods] {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, PackageIntl> = {};
    const methods: PackageContextMethods = {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                return createIntl({
                    locale,
                    messages: packageMessages
                });
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
    return [locale, methods];
}
