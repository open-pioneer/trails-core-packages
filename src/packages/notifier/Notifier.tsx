// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    Toaster as ChakraToaster,
    Icon,
    Portal,
    Spinner,
    Stack,
    Toast,
    useToastStyles
} from "@chakra-ui/react";
import { useIntl, useService } from "open-pioneer:react-hooks";
import { memo, useEffect, useState } from "react";
import { FiAlertCircle, FiAlertTriangle, FiCheckCircle, FiInfo } from "react-icons/fi";
import { InternalNotificationAPI, ToasterObject } from "./NotificationServiceImpl";

/** Props supported by the {@link Notifier} component. */
export interface NotifierProps {}

/**
 * Shows notifications sent via the `NotificationService`.
 *
 * Only one instance of `<Notifier />` should be present in the application.
 * It currently does not matter where the Notifier is located in the react tree.
 *
 * ```ts
 * import { Notifier } from "@open-pioneer/notifier";
 *
 * export function AppUI() {
 *     return (
 *         <>
 *             <Notifier />
 *             <h1>Your application</h1>
 *         </>
 *     );
 * }
 * ```
 */
export function Notifier(_props: NotifierProps) {
    const notificationService = useService(
        "notifier.NotificationService"
    ) as InternalNotificationAPI;
    const [active, setActive] = useState(false);

    useEffect(() => {
        const handle = notificationService.registerUI();
        if (!handle) {
            return;
        }

        setActive(true);
        return () => {
            setActive(false);
            handle.destroy();
        };
    }, [notificationService]);

    return active && <Toaster toaster={notificationService.toaster} />;
}

const Toaster = memo(function Toaster(props: { toaster: ToasterObject }) {
    const intl = useIntl();
    return (
        <Portal>
            <ChakraToaster
                toaster={props.toaster}
                insetInline={{ mdDown: "4" }}
                aria-label={intl.formatMessage({
                    id: "regionLabel"
                })}
            >
                {(toast) => (
                    <Toast.Root width={{ md: "sm" }} alignItems="center">
                        {toast.type === "loading" ? (
                            <Spinner size="sm" color="blue.solid" />
                        ) : (
                            <ToastIndicator type={toast.type ?? ""} />
                        )}
                        <Stack gap="1" flex="1" maxWidth="100%">
                            {toast.title && <Toast.Title>{toast.title}</Toast.Title>}
                            {toast.description && (
                                <Toast.Description>{toast.description}</Toast.Description>
                            )}
                        </Stack>
                        {toast.action && (
                            <Toast.ActionTrigger>{toast.action.label}</Toast.ActionTrigger>
                        )}
                        {toast.meta?.closable && (
                            <Toast.CloseTrigger
                                cursor="pointer"
                                aria-label={intl.formatMessage({
                                    id: "notification.close"
                                })}
                            />
                        )}
                    </Toast.Root>
                )}
            </ChakraToaster>
        </Portal>
    );
});

const icons: Record<string, React.ElementType> = {
    info: FiInfo,
    success: FiCheckCircle,
    warning: FiAlertCircle,
    error: FiAlertTriangle
};

function ToastIndicator(props: { type: string }) {
    const styles = useToastStyles();

    const Component = icons[props.type];
    if (!Component) return null;

    return (
        <Icon css={styles.indicator}>
            <span>
                <FiAlertTriangle style={{ width: "100%", height: "100%" }} />
            </span>
        </Icon>
    );
}
