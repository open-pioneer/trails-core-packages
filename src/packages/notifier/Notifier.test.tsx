// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { ApplicationContext } from "@open-pioneer/runtime";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { act, render, screen, waitFor, waitForElementToBeRemoved } from "@testing-library/react";
import { createElement } from "react";
import { expect, it } from "vitest";
import { NotificationServiceImpl } from "./NotificationServiceImpl";
import { Notifier } from "./Notifier";

it("shows notifications as toasts", async () => {
    const { service, content } = await create();
    render(content);

    act(() => {
        service.notify({
            title: "test",
            // XXX: Cannot use jsx syntax here because it leads to an infinite recursion in zag-js??
            //      Only happens during testing.
            message: createElement(
                "span",
                {
                    "data-testid": "notification-message"
                },
                "Message"
            )
        });
    });

    const messageDiv = await screen.findByTestId("notification-message", {});
    const toastElement = findToastRoot(messageDiv);

    expect(toastElement).toMatchSnapshot();
});

it("closes all notifications", async () => {
    const { service, content } = await create();
    render(content);

    act(() => {
        service.notify({ title: "test1" });
        service.notify({ title: "test1" });
        service.notify({ title: "test1" });
    });

    const messageElements = await waitFor(async () => {
        const messageElements = await screen.findAllByText("test1");
        if (messageElements.length !== 3) {
            throw new Error("Did not open three notifications");
        }
        return messageElements;
    });
    expect(messageElements).toHaveLength(3);

    act(() => {
        service.closeAll();
    });
    await waitForElementToBeRemoved(messageElements);
    const messageElementsAfterClear = await screen.queryAllByText("test1");
    expect(messageElementsAfterClear).toHaveLength(0);
});

async function create() {
    const service = await createService(NotificationServiceImpl, {
        references: {
            appCtx: {
                getShadowRoot() {
                    return document as any;
                }
            } satisfies Partial<ApplicationContext>
        }
    });
    const services = {
        "notifier.NotificationService": service
    };
    const content = (
        <PackageContextProvider services={services}>
            <Notifier />
        </PackageContextProvider>
    );
    return { service, content };
}

function findToastRoot(messageDiv: HTMLElement) {
    let currentElement = messageDiv.parentElement;
    while (currentElement) {
        if (currentElement.classList.contains("chakra-toast__root")) {
            return currentElement;
        } else {
            currentElement = currentElement.parentElement;
        }
    }
    return undefined;
}
