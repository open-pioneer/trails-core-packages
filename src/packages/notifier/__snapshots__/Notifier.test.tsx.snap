// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shows notifications as toasts 1`] = `
<div
  aria-atomic="true"
  aria-describedby="toast:toast:1:description"
  aria-labelledby="toast:toast:1:title"
  class="chakra-toast__root css-1bm8y9v"
  data-align="end"
  data-first=""
  data-mounted=""
  data-part="root"
  data-placement="top-end"
  data-scope="toast"
  data-side="top"
  data-stack=""
  data-state="open"
  data-type="info"
  dir="ltr"
  id="toast:toast:1"
  role="status"
  style="position: absolute; pointer-events: auto; --opacity: 1; --remove-delay: 200ms; --duration: 604800000ms; --initial-height: 0px; --offset: 0px; --index: 0; --z-index: 1; --lift-amount: calc(var(--lift) * var(--gap)); --y: calc(var(--lift) * var(--offset)); --x: 0; top: 0px; --sign: -1; --lift: 1; --height: var(--initial-height);"
  tabindex="0"
>
  <div
    data-ghost="before"
    style="position: absolute; pointer-events: none;"
  />
  <span
    aria-hidden="true"
    class="chakra-icon css-1yhex1d"
    focusable="false"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="width: 100%; height: 100%;"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="16"
        y2="12"
      />
      <line
        x1="12"
        x2="12.01"
        y1="8"
        y2="8"
      />
    </svg>
  </span>
  <div
    class="chakra-stack css-403sr0"
  >
    <div
      class="chakra-toast__title css-1nhm9ic"
      data-part="title"
      data-scope="toast"
      id="toast:toast:1:title"
    >
      test
    </div>
    <div
      class="chakra-toast__description css-1tl80zq"
      data-part="description"
      data-scope="toast"
      id="toast:toast:1:description"
    >
      <span
        data-testid="notification-message"
      >
        Message
      </span>
    </div>
  </div>
  <button
    aria-label="notification.close"
    class="chakra-toast__closeTrigger css-s29zxt"
    data-part="close-trigger"
    data-scope="toast"
    id="toasttoast:1:close"
    type="button"
  >
    <svg
      class="css-kqzqgg"
      viewBox="0 0 24 24"
    >
      <path
        clip-rule="evenodd"
        d="M18.7071 6.70711C19.0976 6.31658 19.0976 5.68342 18.7071 5.29289C18.3166 4.90237 17.6834 4.90237 17.2929 5.29289L12 10.5858L6.70711 5.29289C6.31658 4.90237 5.68342 4.90237 5.29289 5.29289C4.90237 5.68342 4.90237 6.31658 5.29289 6.70711L10.5858 12L5.29289 17.2929C4.90237 17.6834 4.90237 18.3166 5.29289 18.7071C5.68342 19.0976 6.31658 19.0976 6.70711 18.7071L12 13.4142L17.2929 18.7071C17.6834 19.0976 18.3166 19.0976 18.7071 18.7071C19.0976 18.3166 19.0976 17.6834 18.7071 17.2929L13.4142 12L18.7071 6.70711Z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
  <div
    data-ghost="after"
    style="position: absolute; left: 0px; bottom: 100%; width: 100%;"
  />
</div>
`;
